torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+    if (py::isinstance<py::function>(guard_check_fn) &&
+        py::isinstance<py::function>(print_failure_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+      _guard_check_fn_pyobj = _guard_check_fn.ptr();
+      _print_failure_fn = py::cast<py::function>(print_failure_fn);
+    } else {
+      throw py::type_error("PythonLambdaGuard expects callables");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return PyObject_IsTrue(PyObject_CallOneArg(_guard_check_fn_pyobj, value));
+  }
+
+  std::string get_failure_reason(PyObject* value) override {
+    return py::cast<std::string>(_print_failure_fn(py::handle(value)));
+  }
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
+  PyObject* _guard_check_fn_pyobj;
+  // The user provided lambda function to get guard failure reason.
+  py::function _print_failure_fn;
+};
+
+/**
+ * This is an example of relational guard and checks tensor X is not tensor Y.
+ * We construct one leaf guard and install it at as a leaf of two guard managers
+ * (one for X and another for Y). Therefore, this guard is run twice. In the
+ * first invocation, it saves the first value and returns True. In the second
+ * invocation, it compares the saved value with the new value and returns True
+ * if they do not alias.
+ */
+class NoTensorAliasingGuard : public LeafGuard {
+ public:
+  NoTensorAliasingGuard() : _is_first_call(true) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    if (_is_first_call) {
+      _first_tensor = value;
+      _is_first_call = false;
+      return true;
+    }
+    _is_first_call = true;
+    return _first_tensor != value;
+  }
+
+  std::string get_failure_reason(PyObject* value) override { // borrowed ref
+    return "";
+  }
+
+ private:
+  bool _is_first_call;
+  PyObject* _first_tensor;
+};
+
+class GuardManager;
+/**
+ * Base class representing a pair of accessor and the associated guard manager.
+ * The accessor defines how to access the child value from the py::object given
+ * to the parent check function.
+ *
+ * GuardAccessors can be considered equivalent to name() method of Source
+ * objects in guards.py. In python, name() method returns a str which we can
+ * then eval in f_locals and f_globals to retrieve the actual py object.
+ * GuardAccessor serves the same purpose. The minor difference is that
+ * GuardManager is a tree structure, so a GuardAccessor just has to retrieve the
+ * value in the next level in this tree and pass it to the child GuardAccessor.
+ *
+ * GuardAccessor also owns the GuardManager associated with the retrieved value
+ * from the GuardAccessor.
+ */
+class GuardAccessor {
+ public:
+  GuardAccessor(py::object accessor_key)
+      : _guard_manager(std::make_unique<GuardManager>()),
+        _accessor_key(accessor_key) {}
```

Comment:
```suggestion
        _accessor_key(std::move(accessor_key)) {}
```
to avoid an extra incref/decref.